src/fat_beagle.cpp

```diff
@@ -79,24 +79,43 @@ double FatBeagle::LogLikelihood(const RootedTree &tree) const {
   return LogLikelihoodInternals(tree.Topology(), branch_lengths);
 }
 
+// Build differential matrix for gradient wrt branch lengths.
+EigenMatrixXd GetBranchDifferentialMatrices(SiteModel &site_model,
+                                            SubstitutionModel &substitution_model) {
+  size_t category_count = site_model.GetCategoryCount();
+  EigenMatrixXd Q = substitution_model.GetQMatrix();
+  Eigen::Map<Eigen::RowVectorXd> mapQ(Q.data(), Q.size());
+  EigenMatrixXd dQ = mapQ.replicate(category_count, 1);
+  const EigenVectorXd &category_rates = site_model.GetCategoryRates();
+  for (size_t k = 0; k < category_count; k++) {
+    dQ.row(k) *= category_rates[k];
+  }
+  return dQ;
+}
+
+// Build differential matrix for site model gradient.
+EigenMatrixXd GetSiteModelDifferentialMatrices(SiteModel &site_model,
```

Comment:
I have a slight preference for `Build` vs `Get` as the latter sounds like we are just pulling out something that's already there.